exec/cnex/lib/sqlite.c

```diff
@@ -0,0 +1,417 @@
+    return 0;
+}
+
+
+
+void sqlite_open(TExecutor *exec)
+{
+    char *name = string_asCString(top(exec->stack)->string); pop(exec->stack);
+
+    sqlite3 *db;
+    int r = sqlite3_open(name, &db);
+    if (r != SQLITE_OK) {
+        exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db), number_from_sint64(r));
+        free(name);
+        return;
+    }
+
+    push(exec->stack, cell_fromObject(object_createDatabaseObject(db)));
+    free(name);
+}
+
+void sqlite_exec(TExecutor *exec)
+{
+    Cell *parameters = cell_fromCell(top(exec->stack)); pop(exec->stack);
+    char *sql = string_asCString(top(exec->stack)->string); pop(exec->stack);
+    Cell *pdb = cell_fromCell(top(exec->stack)); pop(exec->stack);
+
+    sqlite3_stmt *stmt;
+    Object *db = check_database(exec, pdb->object);
+    if (db == NULL) {
+        goto cleanup;
+    }
+    Cell *rows = cell_createArrayCell(0);
+    int r = sqlite3_prepare_v2(db->ptr, sql, -1, &stmt, NULL);
+    if (r != SQLITE_OK) {
+        exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+        goto cleanup;
+    }
+    for (int i = 0; i < parameters->dictionary->len; i++) {
+        int c = sqlite3_bind_parameter_index(stmt, TCSTR(parameters->dictionary->data[i].key));
+        if (c == 0) {
+            exec->rtl_raise(exec, "SqlException.ParameterName", TCSTR(parameters->dictionary->data[i].key), BID_ZERO);
+            goto cleanup;
+        }
+        r = sqlite3_bind_text(stmt, c, TCSTR(parameters->dictionary->data[i].value->string), -1, SQLITE_TRANSIENT);
+        if (r != SQLITE_OK) {
+            exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+            goto cleanup;
+        }
+    }
+    int columns = sqlite3_column_count(stmt);
+    for (;;) {
+        r = sqlite3_step(stmt);
+        if (r == SQLITE_DONE) {
+            break;
+        }
+        if (r == SQLITE_ROW) {
+            Cell *row = cell_createArrayCell(0);
+            for (int i = 0; i < columns; i++) {
+                cell_arrayAppendElementPointer(row, cell_fromCString((const char *)sqlite3_column_text(stmt, i)));
+            }
+            cell_arrayAppendElementPointer(rows, row);
+        } else {
+            exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+            goto cleanup;
+        }
+    }
+    push(exec->stack, rows);
+
+cleanup:
+    sqlite3_finalize(stmt);
+    free(sql);
+    cell_freeCell(parameters);
+    cell_freeCell(pdb);
+}
+
+void sqlite_execOne(TExecutor *exec)
+{
+    Cell *parameters = cell_fromCell(top(exec->stack)); pop(exec->stack);
+    char *sql = string_asCString(top(exec->stack)->string); pop(exec->stack);
+    Cell *pdb = cell_fromCell(top(exec->stack)); pop(exec->stack);
+
+    Object *db = check_database(exec, pdb->object);
+    if (db == NULL) {
+        cell_freeCell(parameters);
+        free(sql);
+        cell_freCell(pdb);
```

Comment:
This can't possibly have worked?

Reply:
Woah.  Now THAT was a good one!  I must have hit it after compiling/testing, but before adding everything to git.  Fixed.